src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/RecursiveSpinLockTest.kt

```diff
@@ -0,0 +1,502 @@
+        .checkLincheckOutput("spin_lock/recursive_spin_cycle_inner_events.txt")
+
+}
+
+/**
+ * Checks proper output in case of recursive spin-lock in one thread.
+ * Should correctly detect spin cycle and place the spin cycle labels when
+ * the recursion includes two different method calls.
+ */
+class RecursiveSpinLockTwoStepRecursionEventsTest {
+
+    private val counter = AtomicInteger(0)
+    private val someUselessSharedState = AtomicBoolean(false)
+
+    @Operation
+    fun trigger() {
+        counter.incrementAndGet()
+        counter.decrementAndGet()
+    }
+
+    @Operation
+    fun causesSpinLock() {
+        if (counter.get() != 0) {
+            outerRecursiveSpinCycle()
+        }
+    }
+
+    private fun outerRecursiveSpinCycle() {
+        deadSpinCycleRecursive()
+    }
+
+    private fun deadSpinCycleRecursive() {
+        repeat(4) {
+            val value = getSharedVariable()
+            action(value)
+        }
+        outerRecursiveSpinCycle()
+    }
+
+    private fun action(value: Boolean) = someUselessSharedState.compareAndSet(value, !value)
+
+    private fun getSharedVariable(): Boolean = someUselessSharedState.get()
+
+    @Test
+    fun test() = ModelCheckingOptions()
+        .addCustomScenario {
+            parallel {
+                thread { actor(RecursiveSpinLockTwoStepRecursionEventsTest::trigger) }
+                thread { actor(RecursiveSpinLockTwoStepRecursionEventsTest::causesSpinLock) }
+            }
+        }
+        .minimizeFailedScenario(false)
+        .checkImpl(this::class.java)
+        .checkLincheckOutput("spin_lock/recursive_spin_cycle_two_step.txt")
+
+}
+
+/**
+ * Checks proper output in case of recursive spin-lock in two threads.
+ */
+class RecursiveTwoThreadsSpinLockTest {
+    private val sharedState1 = AtomicBoolean(false)
+    private val sharedState2 = AtomicBoolean(false)
+
+    @Operation
+    fun one(): Int {
+        meaninglessActions2()
+        meaninglessActions1()
+
+        sharedState1.set(false)
+        sharedState2.set(false)
+
+        return 1
+    }
+
+    @Operation
+    fun two(): Int {
+        meaninglessActions1()
+        meaninglessActions2()
+
+        sharedState2.set(false)
+        sharedState1.set(false)
+
+        return 2
+    }
+
+    private fun meaninglessActions1() {
+        if (!sharedState2.compareAndSet(false, true)) {
+            meaninglessActions1()
+        }
+    }
+    private fun meaninglessActions2() {
+        if (!sharedState1.compareAndSet(false, true)) {
+            meaninglessActions2()
+        }
+    }
+
+    @Test
+    fun testWithModelCheckingStrategy() = ModelCheckingOptions()
+        .minimizeFailedScenario(true)
+        .checkImpl(this::class.java)
+        .checkLincheckOutput("spin_lock/recursive_two_threads_spin_lock.txt")
+}
+
+
+/**
+ * Checks recursive spin lock representation when execution hung due to alternation of two threads
+ * in recursive live-lock.
+ */
+class BrokenCas2RecursiveLiveLockTest {
```

Comment:
Let's rename into `AtomicArrayWithCAS2LiveLockTest` and together with the `AtomicArrayWithCAS2` class extract it into separate file.